Classes/MapEditor/MapEditor.cs

```diff
@@ -355,6 +357,55 @@ public class MapEditor {
         }
         AddNotes(notes);
     }
+
+    public void QuantizeSelection() {
+        if (currentMapDifficulty == null) {
+            return;
+        }
+        
+        double defaultBeats = 60.0 / (globalBPM * (defaultGridDivision / 2));
+
+        List<Note> notesToAdd = new List<Note>();
+        List<Note> notesToRemove = new List<Note>();
+
+
+        foreach (Note n in currentMapDifficulty.selectedNotes) {
+            BPMChange? currentBeat = currentMapDifficulty
+                .bpmChanges
+                .OrderByDescending(obj => obj.globalBeat)
+                .Where(obj => obj.globalBeat <= n.beat)
+                .Select(obj => obj)
+                .FirstOrDefault();
+            
+            double offset = 0.0;
+
+            // beat has been changed
+            if (currentBeat != null){
+
+                // use current values for defaultBeats calculation
+                defaultBeats = 60.0 / (currentBeat.BPM * (currentBeat.gridDivision / 2));
+
+                // calculate time difference between old beat and start time
+                double differenceDefaultNew = Math.Floor(currentBeat.globalBeat / defaultBeats) * defaultBeats;
+
+                // calculate offset
+                offset = currentBeat.globalBeat - differenceDefaultNew;
+
+            }
+            double newBeat = Math.Round(n.beat / defaultBeats) * defaultBeats;
+            newBeat += offset;
+
+            if (newBeat != n.beat) {
+                Note newNote = new Note(newBeat, n.col);
+                notesToAdd.Add(newNote);
+                notesToRemove.Add(n);
+            }
+        }
+
+        AddNotes(notesToAdd);            
+        RemoveNotes(notesToRemove);
```

Comment:
this still feels wrong somehow shouldn't the objects be handled as mutable reference?